include/cantera/numerics/PreconditionerBase.h

```diff
@@ -0,0 +1,158 @@
+// copyright information.
+
+#ifndef PRECONDITIONERBASE_H
+#define PRECONDITIONERBASE_H
+
+#include "cantera/base/ctexceptions.h"
+#include "cantera/numerics/Integrator.h"
+
+namespace Cantera
+{
+
+//! Forward declarations
+class ReactorNet;
+class Reactor;
+class MoleReactor;
+class IdealGasMoleReactor;
+class IdealGasConstPressureMoleReactor;
+
+//! Flag to indicate preconditioner is not set
+const int PRECONDITIONER_NOT_SET = 0;
+
+//! PreconditionerBase serves as an abstract type to extend different
+//! preconditioners from.
+class PreconditionerBase
+{
+public:
+    PreconditionerBase(/* args */){}
+    ~PreconditionerBase(){}
+
+    //! Use this function to return zero for preconditioner not set
+    virtual size_t getPreconditionerMethod(){return PRECONDITIONER_NOT_SET;};
+
+    //! Use this function to the "type" for CVODES
+    virtual PreconditionerType getPreconditionerType(){return NO_PRECONDITION;};
+
+    //! Use this function to solve a linear system Ax=b where A is the
+    //! preconditioner contained in this matrix
+    //! @param[in] state_len length of the rhs and output vectors
+    //! @param[in] rhs_vector right hand side vector used in linear
+    //! system
+    //! @param[out] output guess vector used by GMRES
+    virtual void solve(const size_t state_len, double *rhs_vector, double* output){
+        throw NotImplementedError("PreconditionerBase::solve");
+    };
+
+    //! Use this function to perform preconditioner specific post-reactor
+    //! setup operations such as factorize.
+    virtual void setup()
+    {
+        throw NotImplementedError("PreconditionerBase::setup");
+    };
+
+    //! This function performs preconditioner specific post-reactor
+    //! setup operations such as factorize.
+    virtual void reset()
+    {
+        throw NotImplementedError("PreconditionerBase::reset");
+    };
+
+    //! This function is called during setup for any processes that
+    //! need to be completed prior to setup functions
+    //! @param network A pointer to the reactor net object
+    //! associated with the integration
+    virtual void initialize(ReactorNet& network){
+        throw NotImplementedError("PreconditionerBase::initialize");
+    };
+
+    //! Use this function to print preconditioner contents
+    virtual void printPreconditioner(){
+        throw NotImplementedError("PreconditionerBase::printPreconditioner");
+    };
+
+    //! This function is for a visitor design pattern to determine
+    //! preconditioner type with Reactor
+    virtual void acceptReactor(Reactor& reactor, double t, double* LHS, double* RHS){
+        throw NotImplementedError("PreconditionerBase::acceptReactor");
+    };
+
+    //! This function is for a visitor design pattern to determine
+    //! preconditioner type with MoleReactor
+    virtual void acceptReactor(MoleReactor& reactor, double t, double* LHS, double* RHS){
+        throw NotImplementedError("PreconditionerBase::acceptReactor");
+    };
```

Comment:
@speth I agree that there is probably room for simplification here but I am concerned that only implemented `void preconditionerSetup(PreconditionerBase& preconditioner, double t, double* LHS, double* RHS)` will not be enough to capture different types of preconditioning. For example, in another branch of mine I have a fully analytical version of the `AdaptivePreconditioner` which uses an analytical temperature derivative. The same `preconditionerSetup` function will not work for both of these so I believe recognizing the specific type of preconditioner is important. Do you have an alternative suggestion to how I can achieve this in a simpler manner?

Reply:
I think this comes down to a question of what behaviors are the responsibility of the `Preconditioner` class and which are part of the `Reactor`. The two main things that have to happen are (a) calculating a [approximate] Jacobian, (b) forming a preconditioner matrix and factorizing it, and (c) Using this factorized matrix to solve linear systems as part of the ODE integration. Since the Jacobian depends on the reactor's governing equations, I think that pretty squarely belongs as part of the `Reactor` implementation, while steps (b) and (c) would fall to the Preconditioner class. This is essentially what the current implementation looks like.

I would say that a different way of calculating the Jacobian elements such as analytical derivatives is something that would belong to the `Reactor` class, not to a different `Preconditioner` class, since you're going to need a different implementation for every `Reactor` variant anyway. The cases I can see where you would want a different preconditioner class would be ones where you either have a different algorithm for picking  which elements to keep and discard, or you want to do the factorization and solve differently, for example using SuperLU or MAGMA.

Reply:
Fwiw, I’m strongly in favor of simplifying the structure here as well. I don’t think that it’s necessary for a `Reactor` implementation to know how to form the preconditioner (same for a `ReactorNet`). Imho, the intuitive way would be to let the `Reactor` handle the Jacobian (part (a) of @speth’s argument above, which by itself may be of interest to the user base), and let the `Preconditioner` be something that is only relevant to the solver (parts (b/c) above)? See also one of my old comments:

> […] I believe this would allow for a very clean hierarchy, i.e.
>
> `AdaptivePreconditioner` -> `ReactorNet` -> `Reactor` -> `Kinetics`

At this point, `Kinetics` knows how to calculate (semi-)analytical derivatives, which is a capability that exists entirely outside preconditioning … I believe that, ideally, the Jacobian evaluation should follow the same philosophy.